internal/olm/client/client.go

```diff
@@ -226,47 +258,60 @@ func (c Client) DoCSVWait(ctx context.Context, key types.NamespacedName) error {
 
 	err := wait.PollImmediateUntil(time.Second, csvPhaseSucceeded, ctx.Done())
 	if err != nil && errors.Is(err, context.DeadlineExceeded) {
-		if depCheckErr := c.printDeploymentErrors(ctx, key, csv); depCheckErr != nil {
-			return fmt.Errorf("error printing operator resource errors: %v %v", err, depCheckErr)
+		depCheckErr := c.checkDeploymentErrors(ctx, key, csv)
+		if depCheckErr != nil {
+			return depCheckErr
 		}
 	}
 	return err
 }
 
-// TODO(btenneti) Refactor function to collect errors into customized error and return.
-// printDeploymentErrors function loops through deployment specs of a given CSV, and prints reason
+// checkDeploymentErrors function loops through deployment specs of a given CSV, and prints reason
 // in case of failures, based on deployment condition.
-func (c Client) printDeploymentErrors(ctx context.Context, key types.NamespacedName, csv olmapiv1alpha1.ClusterServiceVersion) error {
+func (c Client) checkDeploymentErrors(ctx context.Context, key types.NamespacedName, csv olmapiv1alpha1.ClusterServiceVersion) error {
+	depErr := deploymentErrors{}
 	if key.Namespace == "" {
 		return fmt.Errorf("no namespace provided to get deployment failures")
 	}
 	dep := &appsv1.Deployment{}
 	for _, ds := range csv.Spec.InstallStrategy.StrategySpec.DeploymentSpecs {
+		podErr := podErrors{}
 		depKey := types.NamespacedName{
 			Namespace: key.Namespace,
 			Name:      ds.Name,
 		}
 		depSelectors := ds.Spec.Selector
 		if err := c.KubeClient.Get(ctx, depKey, dep); err != nil {
-			log.Printf("error getting operator deployment %q: %v", ds.Name, err)
+			depErr = append(depErr, deploymentError{
+				resourceError{
+					name:  ds.Name,
+					issue: err.Error(),
+				},
+				podErr,
+			})
 			continue
 		}
 		for _, s := range dep.Status.Conditions {
-			if s.Type == appsv1.DeploymentAvailable && s.Status == corev1.ConditionFalse {
-				log.Printf("operator deployment %q not available: %s", ds.Name, s.Reason)
-				if err := c.printPodErrors(ctx, depSelectors, key); err != nil {
-					return err
-				}
+			if s.Type == appsv1.DeploymentAvailable && s.Status != corev1.ConditionTrue {
+				podError := c.checkPodErrors(ctx, depSelectors, key)
+				podErr = podError.(podErrors)
+				depErr = append(depErr, deploymentError{
+					resourceError{
+						name:  ds.Name,
+						issue: s.Reason,
+					},
+					podErr,
+				})
```

Comment:
```suggestion
				depErr := deploymentError{
					resourceError{
						name:  ds.Name,
						issue: s.Reason,
					},
				})
				podErr := c.checkPodErrors(ctx, depSelectors, key)
				podErrs := &podErrors{}
				if errors.As(podErr, &podErrs) {
					depErr.podErrs = append(depErr.podErrs, podErrs...)
				} else {
					return podErr
				}
				depErrs = append(depErrs, depErr)
```

Reply:
Can you elaborate on why this change is required?
```go
				podError := c.checkPodErrors(ctx, depSelectors, key)
				podErr = podError.(podErrors)
				depErr = append(depErr, deploymentError{
					resourceError{
						name:  ds.Name,
						issue: s.Reason,
					},
					podErr,
				})
```
Current logic seems to be achieving the same result.

Reply:
If you do a naked type cast, you will get a panic when `err` is not of that type.

Reply:
```go
	if errors.As(podErr, &podErrs) {
					depErr.podErrs = append(depErr.podErrs, podErrs...)
```
This logic doesnt append the actual `podErr` returned in Line 382 above

Reply:
`podErrs` had to be dereferenced here